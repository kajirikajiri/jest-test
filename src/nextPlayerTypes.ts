export type NextPlayer = "X" | "O" | "";
