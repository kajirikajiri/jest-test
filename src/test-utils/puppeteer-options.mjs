export const headless = {
  executablePath: "/usr/bin/chromium",
  headless: true,
  ignoreDefaultArgs: ["--disable-extensions"],
  args: [
    "--disable-gpu",
    "--disable-dev-shm-usage",
    "--disable-setuid-sandbox",
    "--no-first-run",
    "--no-sandbox",
    "--no-zygote",
    "--single-process",
  ],
};

export const nonHeadless = {
  executablePath: "/usr/bin/google-chrome",
  headless: false,
};
