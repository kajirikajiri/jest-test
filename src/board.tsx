import React, { useState } from "react";
import { Square } from "./square";
import { NextPlayer } from "./nextPlayerTypes";

type BoardProps = {
  squares: NextPlayer[];
  updateHistory: (currentHistory: NextPlayer[]) => void;
  changingHistory: boolean;
  xIsNext: boolean;
  win: boolean;
};

export const Board: React.FC<BoardProps> = ({
  xIsNext,
  updateHistory,
  squares,
  changingHistory,
  win,
}) => {
  const nextPlayer = (xIsNext: boolean): NextPlayer => (xIsNext ? "X" : "O");
  const winner = (xIsNext: boolean): NextPlayer => (xIsNext ? "O" : "X");

  const handleClick = (i: number): void => {
    if (!changingHistory && (squares[i] !== "" || win)) return;

    const newSquares: NextPlayer[] = squares.slice();
    newSquares[i] = nextPlayer(xIsNext);
    updateHistory(newSquares);
  };

  const renderSquare = (i: number) => {
    return <Square onClick={() => handleClick(i)} value={squares[i]} />;
  };

  const status: String = win
    ? "winner: " + winner(xIsNext)
    : "Next player: " + nextPlayer(xIsNext);

  return (
    <div>
      <div className="status" data-test="gameStatus">
        {status}
      </div>
      <div className="board-row">
        {renderSquare(0)}
        {renderSquare(1)}
        {renderSquare(2)}
      </div>
      <div className="board-row">
        {renderSquare(3)}
        {renderSquare(4)}
        {renderSquare(5)}
      </div>
      <div className="board-row">
        {renderSquare(6)}
        {renderSquare(7)}
        {renderSquare(8)}
      </div>
    </div>
  );
};
