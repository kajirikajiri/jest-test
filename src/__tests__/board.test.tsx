import React from "react";
import { Board } from "../board";
import renderer from "react-test-renderer";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { NextPlayer } from "../nextPlayerTypes";

let container!: any;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

test("Board class", () => {
  const component = renderer.create(
    <Board
      squares={[""]}
      updateHistory={() => {}}
      changingHistory={false}
      win={false}
      xIsNext={true}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("win: true, xIsNext: true -> winner: O", () => {
  act(() => {
    ReactDOM.render(
      <Board
        squares={[""]}
        updateHistory={() => {}}
        changingHistory={false}
        win={true}
        xIsNext={true}
      />,
      container
    );
  });
  const status = container.querySelector("[data-test=gameStatus]");
  expect(status.innerHTML).toEqual("winner: O");
});

test("win: true, xIsNext: false -> winner: X", () => {
  act(() => {
    ReactDOM.render(
      <Board
        squares={[""]}
        updateHistory={() => {}}
        changingHistory={false}
        win={true}
        xIsNext={false}
      />,
      container
    );
  });
  const status = container.querySelector("[data-test=gameStatus]");
  expect(status.innerHTML).toEqual("winner: X");
});

test("win: false, xIsNext: true -> Next player: X", () => {
  act(() => {
    ReactDOM.render(
      <Board
        squares={[""]}
        updateHistory={() => {}}
        changingHistory={false}
        win={false}
        xIsNext={true}
      />,
      container
    );
  });
  const status = container.querySelector("[data-test=gameStatus]");
  expect(status.innerHTML).toEqual("Next player: X");
});

test("win: false, xIsNext: false -> Next player: O", () => {
  act(() => {
    ReactDOM.render(
      <Board
        squares={[""]}
        updateHistory={() => {}}
        changingHistory={false}
        win={false}
        xIsNext={false}
      />,
      container
    );
  });
  const status = container.querySelector("[data-test=gameStatus]");
  expect(status.innerHTML).toEqual("Next player: O");
});

// test("click square button element then change innerHTML blank to X, next change blank to O", () => {
//   act(() => {
//     ReactDOM.render(<Board />, container);
//   });
//   const button = container.querySelectorAll("[data-test=square]");
//   act(() => {
//     button[0].dispatchEvent(new MouseEvent("click", { bubbles: true }));
//   });
//   expect(button[0].innerHTML).toEqual("X");
//   act(() => {
//     button[1].dispatchEvent(new MouseEvent("click", { bubbles: true }));
//   });
//   expect(button[1].innerHTML).toEqual("O");
// });

// test("click square button element then change innerHTML blank to X, next click same elements then X not change", () => {
//   act(() => {
//     ReactDOM.render(<Board />, container);
//   });
//   const button = container.querySelectorAll("[data-test=square]");
//   act(() => {
//     button[0].dispatchEvent(new MouseEvent("click", { bubbles: true }));
//   });
//   expect(button[0].innerHTML).toEqual("X");
//   act(() => {
//     button[0].dispatchEvent(new MouseEvent("click", { bubbles: true }));
//   });
//   expect(button[0].innerHTML).toEqual("X");
// });

// test("show winner states", () => {
//   act(() => {
//     ReactDOM.render(<Board />, container);
//   });
//   const button = container.querySelectorAll("[data-test=square]");
//   act(() => {
//     button[0].dispatchEvent(new MouseEvent("click", { bubbles: true }));
//   });
//   act(() => {
//     button[1].dispatchEvent(new MouseEvent("click", { bubbles: true }));
//   });
//   act(() => {
//     button[3].dispatchEvent(new MouseEvent("click", { bubbles: true }));
//   });
//   act(() => {
//     button[2].dispatchEvent(new MouseEvent("click", { bubbles: true }));
//   });
//   act(() => {
//     button[6].dispatchEvent(new MouseEvent("click", { bubbles: true }));
//   });
//   const status = container.querySelector("[data-test=gameStatus]");
//   expect(status.innerHTML).toEqual("winner: X");
// });
