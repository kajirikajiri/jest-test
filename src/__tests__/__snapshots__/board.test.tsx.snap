// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board class 1`] = `
<div>
  <div
    className="status"
    data-test="gameStatus"
  >
    Next player: X
  </div>
  <div
    className="board-row"
  >
    <button
      className="square"
      data-test="square"
      onClick={[Function]}
    >
      
    </button>
    <button
      className="square"
      data-test="square"
      onClick={[Function]}
    />
    <button
      className="square"
      data-test="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="board-row"
  >
    <button
      className="square"
      data-test="square"
      onClick={[Function]}
    />
    <button
      className="square"
      data-test="square"
      onClick={[Function]}
    />
    <button
      className="square"
      data-test="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="board-row"
  >
    <button
      className="square"
      data-test="square"
      onClick={[Function]}
    />
    <button
      className="square"
      data-test="square"
      onClick={[Function]}
    />
    <button
      className="square"
      data-test="square"
      onClick={[Function]}
    />
  </div>
</div>
`;
